File: calc.cpp
#include "calc.h"
using namespace std;

Calculator::Calculator() : output(0)
{
	cout << "\nNew calculator generated.\n\n";
}

Calculator::Calculator(map<string, double> history): output(0), _history(history)
{	

	if (!history.empty())
	{
		auto it = --_history.end();
		infix = it->first;
		output = it->second;

	}
}

Calculator::~Calculator()
{
	_history.clear();
}

double Calculator::computeNew(string expression)
{
	infix = expression;
	postfix = toPostfix(infix);

	/*
	cout << "Infix: " << infix << endl;
	cout << "Expression: " << expression << endl;

	cout << "Postfix: ";
	for (auto it = postfix.begin(); it != postfix.end(); ++it)
		cout << *it << " ";
	cout << endl;
	*/

	if (postfix.empty())
		output = INFINITY; // return very large number to indicate error

	else {
		stack<double> numbers;
		while (!postfix.empty()) {
			string curr = postfix[0];

			bool isNum = curr.size() > 1 || charType(curr[0]) == 1;
			if (isNum)
			{
				double value = stod(curr);
				numbers.push(value);
			}

			else {
				output = numbers.top();
				numbers.pop();

				if (curr[0] == '+')
					output += numbers.top();

				else if (curr[0] == '-')
					output = numbers.top() - output;

				else if (curr[0] == '*')
					output *= numbers.top();

				else if (curr[0] == '/')
					output = numbers.top() / output;

				numbers.pop();
			}

			postfix.erase(postfix.begin());

		}
	}

	_history.insert(pair<string, double>(infix, output));

	return output;
}

vector<string> Calculator::infixHistory()
{
	vector<string> ret;
	for (auto it = _history.begin(); it != _history.end(); ++it)
		ret.push_back(it->first);

	return ret;
}

vector<double> Calculator::valueHistory()
{
	vector<double> ret;
	for (auto it = _history.begin(); it != _history.end(); ++it)
		ret.push_back(it->second);

	return ret;
}

void Calculator::clearHistory()
{
	_history.clear();

}

void Calculator::deleteHistory(int index)
{
	if (index < 0)
		return;

	auto it = _history.begin();
	for (int i = index; i > 0; i--)
		++it;
	
	_history.erase(it);

	if (index == 0) {
		it = --_history.end();
		infix = it->first;
		output = it->second;
	}
}

vector<string> Calculator::toPostfix(string input)	// something off about this
{
	vector<string> output;

	for (auto it = input.begin(); it != input.end(); ++it)
	{
		int validChar = charType(*it);

		if (!validChar)
			return vector<string>();

		if (validChar == 1) {
			string temp = string(1,*it);
			// continued number means a multiple-digit number
			if (!output.empty() && it != input.begin() && charType(*(it-1)) == 1) {
				int number = stoi(output.back()) * 10 + stoi(temp);
				output.back() = to_string(number);
			}
			else {
				output.push_back(temp);
			}
		}

		else if (validChar == 2) {
			op_stack.push(*it);
		}

		else if (validChar == 3) {
			if (op_stack.empty())	// mismatched parentheses
				return vector<string>();

			while (op_stack.top() != '(') {
				if (op_stack.empty())
					return vector<string>();
				
				output.push_back(string(1,op_stack.top()));
				op_stack.pop();
			}

			if (op_stack.top() == '(')
				op_stack.pop();
		}

		else if (validChar >= 4) {
			while (!op_stack.empty() && charType(op_stack.top()) >= validChar) {
				output.push_back(string(1, op_stack.top()));
				op_stack.pop();
			}
			op_stack.push(*it);
		}
		
	}

	while (!op_stack.empty()) {
		if (op_stack.top() != '(')
			output.push_back(string(1, op_stack.top()));
		
		op_stack.pop();
	}

	return output;
}

int Calculator::charType(char letter)
{
	if ('0' <= letter && letter <= '9') // is a number 0-9
		return 1;

	else if (letter == '(')
		return 2;

	else if (letter == ')')
		return 3;

	else if (letter == '+' || letter == '-') // is a precedence 1 operator
		return 4;

	else if (letter == '*' || letter == '/') // is a precedence 2 operator
		return 5;
	
	else
		return 0;	// invalid char
}

File: calc.js
// import stack from "./stack";
const digits = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
const op1 = ["+", "-"];
const op2 = ["×", "÷", "^"];
const allowed = digits.concat(op1, op2);

const toPostfix = (equation) => {
  let output = [];
  let op_stack = [];

  for (let i = 0; i < equation.length; i++) {
    const char = equation[i];
    if (!allowed.includes(char)) {
      return [];
    }

    if (digits.includes(char)) {
      // is a multi-digit number
      if (output.length > 0 && i !== 0 && digits.includes(equation[i - 1])) {
        const new_num = Number(output.pop()) * 10 + Number(char);
        output.push(new_num);
      } else {
        output.push(char);
      }
    } else if (char === "(") {
      op_stack.push(char);
    } else if (char === ")") {
      // mismatched parentheses
      if (op_stack.length === 0) {
        return [];
      }

      while (op_stack.at(-1) !== "(") {
        if (op_stack.length === 0) {
          return [];
        }

        output.push(op_stack.at(-1));
        op_stack.pop();
      }

      if (op_stack.at(-1) === "(") {
        op_stack.pop();
      }
    }

    // operators
    else if (op1.includes(char)) {
      // higher or equal precedence
      const mayCompute = op2.includes(char)
        ? op1.concat(op2).includes(op_stack.at(-1))
        : op1.includes(op_stack.at(-1));

      while (op_stack.length !== 0 && mayCompute) {
        output.push(op_stack.at(-1));
        op_stack.pop();
      }
      op_stack.push(char);
    }
  }

  while (op_stack.length !== 0) {
    if (op_stack.top !== "(") {
      output.push(op_stack.at(-1));
    }
    op_stack.pop();
  }

  return output;
};

const evaluate = (equation) => {
  console.log(`Computing ${equation}...\n`);
  const postfix = toPostfix(equation);
  console.log(`Postfix: ${postfix}`);
  return postfix.length === 0 ? "Invalid input" : 23;
};

console.log(`\nResult: ${evaluate("12-3+1")}`);
